client/lorawan.go

```diff
@@ -4,64 +4,67 @@ import (
 	"bytes"
 	"context"
 	"encoding/json"
-	"errors"
+	"fmt"
 	"net/http"
+	"net/url"
+	"time"
 
 	"github.com/NickGowdy/deveui-cli/device"
 )
 
-// Client used to communicate to external services
-type Client interface {
-	Do(*http.Request) (resp *http.Response, err error)
-}
-
 // LoraWAN used to communicate to LoRaWAN external system
 type LoraWAN struct {
-	baseURL string
-	client  Client
-}
-
-func NewLoraWAN(baseURL string, client Client) *LoraWAN {
-	return &LoraWAN{
-		baseURL: baseURL,
-		client:  client,
-	}
+	fullURL *url.URL
+	client  *http.Client
 }
 
 const endpoint = "/sensor-onboarding-sample" // endpoint for saving DevEUI via LoRaWAN
 
-// RegisterDevice registers new device using LoraWAN external service
-func (l *LoraWAN) RegisterDevice(ctx context.Context) (*device.Device, error) {
-	device := device.NewDevice()
-	identifier := device.GetIdentifier()
-	b := new(bytes.Buffer)
-	reqBody := map[string]string{"Deveui": identifier}
-
-	err := json.NewEncoder(b).Encode(&reqBody)
+func NewLoraWAN(baseURL string, timeout time.Duration) (*LoraWAN, error) {
+	path, err := url.JoinPath(baseURL, endpoint)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to join url: %w", err)
 	}
 
-	fullUrl := l.baseURL + endpoint
-	req, err := http.NewRequestWithContext(ctx, "POST", fullUrl, b)
+	fullURL, err := url.Parse(path)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to parse url: %w", err)
 	}
 
-	resp, err := l.client.Do(req)
+	return &LoraWAN{
+		fullURL: fullURL,
+		client: &http.Client{
+			Timeout: timeout,
+		},
+	}, nil
+}
+
+// RegisterDevice registers a new device using LoraWAN external service
+func (l *LoraWAN) RegisterDevice(ctx context.Context, newDevice *device.Device) error {
```

Comment:
Having the device as a parameter makes this better, because
- the function previously did two things: 1. create a new device 2. register it. Now it has a single responsibility.
- now that we reduced coupling, testing is easier as well